### DISCUSSION TITLE
Read properties of attribute in source generator

### DISCUSSION THREAD

**Discussion by MelGrubb** [dotnet/roslyn#51361] [Q&A]
I'm writing a C# source generator, and I'm marking those classes that it should augment using a custom attribute. I'd like to pass some options on that attribute as well, but I can't find an example of doing anything like that. I know how I would do this via reflection, but an AttributeSyntax is not an Attribute, so none of my previous knowledge applies. I've been poking around at it for a while this morning, but I'm just not seeing it. For this example, let's assume something simple like a simple string that will be appended to the name of the generated class, or a simple switch that controls whether to generate a "full" or a "light" version of the output. How would I read that value from the attribute from within my source generator?

**Comment by VolodymyrBS** [comment]
take a look at AutoNotifyGenerator from generator examples [https://github.com/dotnet/roslyn-sdk/blob/0abb5881b483493b198315c83b4679b6a13a4545/samples/CSharp/SourceGenerators/SourceGeneratorSamples/AutoNotifyGenerator.cs](https://github.com/dotnet/roslyn-sdk/blob/0abb5881b483493b198315c83b4679b6a13a4545/samples/CSharp/SourceGenerators/SourceGeneratorSamples/AutoNotifyGenerator.cs) 
it has an optional attribute parameter with property name. This parameter retrieved 
[at line 119](https://github.com/dotnet/roslyn-sdk/blob/0abb5881b483493b198315c83b4679b6a13a4545/samples/CSharp/SourceGenerators/SourceGeneratorSamples/AutoNotifyGenerator.cs#L119)

**Reply by MelGrubb** [reply]
I’ve taken a look, and I’m probably missing something here. Let’s say I want an attribute that’s on the class itself. I already have the syntax receiver filtering for the classes with the desired attribute like this:

```
        public List<ClassDeclarationSyntax> Classes { get; } = new();

        public void OnVisitSyntaxNode(SyntaxNode syntaxNode)
        {
            if (syntaxNode is ClassDeclarationSyntax @class && @class.AttributeLists.Any(x => x.Attributes.Any(a => a.Name + "Attribute" == nameof(MyAttribute))))
            {
                Classes.Add(@class);
            }
        }
```

Then, in my generator, I’m looping over the classes like this:

```
        public void Execute(GeneratorExecutionContext context)
        {
            var receiver = (BuilderGeneratorSyntaxReceiver)context.SyntaxReceiver!;

            foreach (var @class in receiver.Classes.Where(x => x != null))
            {
                var attribute = @class.AttributeLists.SelectMany(x => x.Attributes).Single(x => x.Name + "Attribute" == nameof(MyAttribute));
                // TODO: Read the properties from the attribute
            …
```
The sample you’ve pointed me to appears to be GENERATING an attribute. I’m wanting to consume the attribute that called my attention to this class in the first place. At this point, I’ve determined that this class is one that I want to generate a partial for, but I’d really like to read some options from the attribute, which already exists in the code.

This is all pretty new to me. I've done similar code generation via T4, CSX, and TypeWriter files. They were more directly based on reflection though, so I was ready the attributes directly from the code and then creating new stuff. I'm trying to translate this to Source Generators, which I think will be a much nicer way to integrate this into our projects.

Any guidance is appreciated. It's hard to look stuff up when I don't yet know what I'm looking for, and the online documentation of the classes is pretty much empty.

**Reply by PathogenDavid** [reply]
> The sample you’ve pointed me to appears to be GENERATING an attribute.

It's definitely reading it. It's getting the `AutoNotify` attribute on line 118:

```csharp
AttributeData attributeData = fieldSymbol.GetAttributes().Single(ad => ad.AttributeClass.Equals(attributeSymbol, SymbolEqualityComparer.Default));
```

And then (optionally) reading the `PropertyName` property of that attribute:

```csharp
TypedConstant overridenNameOpt = attributeData.NamedArguments.SingleOrDefault(kvp => kvp.Key == "PropertyName").Value;
```

Looking at it from the other end, it's looking for attributes like these:

```csharp
public partial class ExampleViewModel
{
    [AutoNotify]
    private string _text = "private field text";

    [AutoNotify(PropertyName = "Count")]
    private int _amount = 5;
}
```

> Any guidance is appreciated. It's hard to look stuff up when I don't yet know what I'm looking for, and the online documentation of the classes is pretty much empty.

It's worth remembering that source generators are basically just Roslyn analyzers with code output. So resources relating to analyzers tend to relate to source generators as well.

You may find @davidwengier's [Source Generator Playground](https://sourcegen.dev/) helpful for tinkering with simple source generators.

[SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzhgHwAEAmARgFgAoIgBgAIBBAVwwgDkIMBLAMwE8A3NWpEAzPQAO2KD2wAbeqXoBRBNgC2k+TABq3GAHcAshAAmMedQDe1evfoBtFm048BAXTsPxSsgwB9DBgEDHoAXnoAIkkobgA3bGD6XgN5M3pg0KjhKm97Z1YOLj5+AAoABSgISRhZfnZNGAjogGEIZgA7DCiASi8qByUJbm76AM0OsciAVlyAXxEqTqbcaTBml2L3fht8p0YMDDjgVhgAVVxsAHMYMsPj7lPggBUZW4xcADoAMTSzND0ACSnQAFnVuMEMpFeApcDBAYx5PIICZmPIeNpmjC4TB+vt4QoYBllFs3KUHiczvQQEwjlTgvtbIMhj4JGSSgJKU8zmVevshszWQ5Fizhb4iP56FUanUMA0mvRrPQPoJ6PCMGrRUNRaKgA===) can show nice visualization for Rosyln's syntax tree. A [similar tool exists within Visual Studio](https://docs.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/syntax-visualizer?tabs=csharp). (Although it's worth keeping in mind that parsing the syntax tree directly is not always the best way to reason about things.)

Finally if you plan on building Roslyn syntax trees directly in your source generator (not my personal preference but I know a lot of people like it), @KirillOsenkov's [Roslyn Quoter](https://roslynquoter.azurewebsites.net/) tool can be very helpful for that.

**Reply by MelGrubb** [reply]
I've traced backwards from that line until I got up to line 48 defining "compilation" by calling "AddSyntaxTrees", which seemed more related to adding new code than reading existing code.

I suppose the more succinct question might be how I navigate from the AttributeSyntax or ClassDeclarationSyntax that I have to AttributeData, similar to line 118 of the sample you pointed me to. I've stopped the debugger at the point I have the AttributeSyntax handy, and tried rooting around in there, looking for anything useful, but I can't find it. This is compounded by the fact that the actual documentation is pretty much empty at this point (for example, https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.csharp.syntax.attributesyntax?view=roslyn-dotnet). There's just nothing there in the way of explanation.

I am not building syntax trees. I'm actually doing some pretty simple text substitution, injecting values into a template, similar to what I did with T4 and CSX templates in the past. I'm sure there are other ways to tackle the problem, and maybe I'll look into them later on, but text templates will do the job for the moment. That's why I need the attribute properties. I'm trying to provide a mechanism for the consuming project to override the built-in templates in my generator. You can see what I'm working on at line 26, here https://github.com/MelGrubb/BuilderGenerator/blob/main/src/BuilderGenerator/BuilderGenerator.cs

I'll look at the other links you've given me and experiment some more, but any guidance here would be helpful. I'm just not seeing the path from what I have to what I need.

**Reply by MelGrubb** [reply]
Debugging some more, it appears that I _can_ actually see the values in the attribute arguments. I've added two named parameters to the attribute, and when I look in attribute.ArgumentList.Arguments, I see two items. Their order is dependent on the positions that the attributes were provided, not on the order they were defined, so I just need to find a proper way to tell them apart. In the debug view, the ToString makes it easy to find the right one, but interrogating the object itself isn't working quite as well. I'm still digging, but so far I haven't found the property I'm looking for that clearly identifies which argument the value is for. I suppose I can just ToString it, and do some string parsing, but that just seems... ew.

**Reply by PathogenDavid** [reply]
> I've traced backwards from that line until I got up to line 48 defining "compilation" by calling "AddSyntaxTrees", which seemed more related to adding new code than reading existing code.

That's because this source generator also injects the actual attribute class you need to use it. (This is more apparent if you look at the output on the source generator playground.)

You don't have to do it that way (and doing it that way has negative performance implications. https://github.com/dotnet/roslyn/issues/49753 fixes that, but it isn't in a release yet.)

> I suppose the more succinct question might be how I navigate from the AttributeSyntax or ClassDeclarationSyntax that I have to AttributeData

Ah, sorry. I overlooked that part.

The ideal way to inspect the values is using the semantic model rather than the syntax tree.

Roughly speaking it'll look something like this:

```csharp
SemanticModel model = context.Compilation.GetSemanticModel(@class.SyntaxTree);
ISymbol classSymbol = model.GetDeclaredSymbol(@class);
INamedTypeSymbol classNamedTypeSymbol = (INamedTypeSymbol)classSymbol; // In a real case you'd want to do some sanity validation here. I emit a diagnostic when GetDeclaredSymbol returns null or something other than a named type.

AttributeData? attributeData = classNamedTypeSymbol.GetAttributes().FirstOrDefault(/*...*/);
```

(The `AutoNotifyGenerator` example [does this ahead of time](https://github.com/dotnet/roslyn-sdk/blob/0abb5881b483493b198315c83b4679b6a13a4545/samples/CSharp/SourceGenerators/SourceGeneratorSamples/AutoNotifyGenerator.cs#L54-L68), hence why it was seemingly missing from the lines we were talking about.)

**Reply by MelGrubb** [reply]
I've given that a try, and I seem to be able to drill into the attribute from this angle as well, but then the constructor arguments are missing. Here is what I have at the moment

```
            foreach (var @class in receiver.Classes.Where(x => x != null))
            {
                var model = context.Compilation.GetSemanticModel(@class.SyntaxTree);
                var classSymbol = model.GetDeclaredSymbol(@class) as INamedTypeSymbol;
                var attributeDatas = classSymbol.GetAttributes();
                var attributeData = attributeDatas.FirstOrDefault(x => $"{x.AttributeClass?.Name}Attribute" == nameof(GenerateBuilderAttribute));
                var constructorArguments = attributeData.ConstructorArguments.ToList();
                var namedArguments = attributeData.NamedArguments.ToList();
               ...
```

I get the attribute back just fine, but both the contructorArguments and namedArguments are coming up empty. For reference, here is what the actual class I'm trying to interrogate currently looks like:

```
    [GenerateBuilder("attributeTemplate", "classTemplate", buildMethodSetterTemplate: "setterTemplate")]
    public class Foo
    {
    ...
```

As you can see, I'm specifying three constructor arguments, two positional and one named. I'm anticipating that the attribute will only even use named arguments to override specific built-in templates, but the possibility still exists that someone could use the positional arguments, so I need to handle both.

I swear I almost had something last night when my brain ran out. In the debugger, looking at the AttributeSyntax-based approach, which seems to get me more information, I still end up with something like this:

```
AttributeArgumentListSyntax AttributeArgumentList ("attributeTemplate", "classTemplate", buildMethodSetterTemplate: "setterTemplate")
```

If I drill into ArgumentList.Arguments, I can see the array of three arguments, but I don't see the names for the first two, only their values. The third argument knows its name AND value. I'd hate to have to count on knowing the argument positions. That seems hard to build on top of later if anything should move around.

If I can manage to extract a general-purpose function that just returns me a dictionary of the attribute's constructor parameters, named and/or positional, then I should be done here. I don't care whether it's using AttributeSyntax or AttributeData. I just can't seem to dig out the things I need from either angle. I look where the data should be, and it's just not there.

I appreciate the help here. I've been digging at this in the evenings for days now, and I'm just not finding it. I'll continue to look at the code examples as well in case something becomes clearer.

**Reply by jasonmalinowski** [reply]
For sake of debugging, if you look at the AttributeData, and you look at the AttributeConstructor member, what is that? Is that a regular constructor or some sort of error fallback?

**Comment by MelGrubb** [comment]
It’s null.

[cid:image003.png@01D7094C.167882C0]

**Reply by MelGrubb** [reply]
Sooo close. I have a method here, working off of AttributeSyntax rather than AttributeData. It works, but only when the constructor parameters have been named.

```
        private string? GetAttributeValue(AttributeSyntax attribute, string name)
        {
            var argument = attribute.ArgumentList?.Arguments.SingleOrDefault(x => x.NameColon?.Name.Identifier.ValueText.Equals(name, StringComparison.InvariantCultureIgnoreCase) == true);
            var expression = argument?.Expression as LiteralExpressionSyntax;
            var value = expression?.Token.ValueText;

            return value;
        }
```

I suppose that's because "NameColon" only exists when using named parameters. Again, I'm flying blind here, so I'm sure I have several subtleties wrong. I think if I could get the actual constructor parameters from the definition of the constructor into some kind of an array, then I could somehow zip the two lists together, combining the named and positional parameters to make one single list out of them. As it is, if you use named parameters for everything, it'll work just fine, but of course tools like ReSharper will remove the names when the order matches the positional arguments, and telling users to shuffle the parameter order around seems pretty terrible.

I'm currently trying to build a similar method that works against AttributeData, but to be honest, I've been seeing better "success" from AttributeSyntax so far.

**Reply by jasonmalinowski** [reply]
> I think if I could get the actual constructor parameters from the definition of the constructor into some kind of an array, then I could somehow zip the two lists together, combining the named and positional parameters to make one single list out of them.

So that's what AttributeData will be able to give you -- the constructor symbol would have the parameters you can line up from there. But if I'm following above, the attribute is being defined by the generator so it's not yet available for binding?

**Reply by MelGrubb** [reply]
No, my attribute is hard-coded into the generator project. It doesn't change. You have to basically reference the generator project twice, once as an analyzer, and once as a library, but the attribute is visible to the generator as well as to the consuming project without me having to break out a separate "Common" assembly. I was pretty pleased when that worked.

**Reply by MelGrubb** [reply]
If you want to see the full source of where I am at the moment, I've pushed it on a separate branch: https://github.com/MelGrubb/BuilderGenerator/tree/Customizable_templates

The part I'm working on starts at line 57 in BuilderGenerator.cs, with the "GetTemplates" method.

**Reply by MelGrubb** [reply]
I spent some time looking at the samples again, and I may just change direction here. The Mustache sample shows a pretty simple way to handle the customization via attributes. I'd rather not have to stack up multiples, but it works, right? The CSV example shows how reading external files would work as well, and I already have that working as of last night. I think a combination of the two would get me everything I need. It should let a consumer provide their own set of templates if they want via the AdditionalFiles route. The files themselves can just be simple text files with no weird escaping other than the tag placeholders. Then, if you wanted to override something more specifically, you could decorate the class with one or more attributes, each overriding a specific template.

It just feels like that's the path of least resistance right now, and maybe I should just go with that, and maybe come back to my original idea later on, if anyone even needs it. I don't see template customization as being very popular. The package makes builders. The builders work. There's a mechanism to change them if you want. Done.
